### PR TITLE
Reworked conditional formatting behavior and added some changes

### PR DESCRIPTION
**Pull request by b0bi79**
FIXED
~ Shifting Conditional Formats
~ Removing Conditional Formats
~ Copying Conditional Formats
~ Reset _dataType for cell when Cell.Clear()
~ Some leaks of Range Shifted Subscriptions.

ADDED
+ RangeExtensions functions Crop and Relative
+ Conditional formats compress function
+ While saving, the conditional formats will be compressed
+ When cleaning a range, the conditional formatting will be removed

### CODE REVIEW

**Comment by igitur** [comment]
Resolve the conflict please. I will look at this soon then.

**Comment by b0bi79** [comment]
Resolved

**Comment by igitur** [comment]
Thanks. Also check out the review comments.

**Comment by b0bi79** [comment]
Sorry, but I do not see any review comments.

**Comment by igitur** [comment]
Do you see them now?

File: ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs [hunk 1]
```diff
@@ -170,6 +170,95 @@ public void WsNamedRangesOneString()
...
+            var ws = new XLWorkbook().Worksheets.Add("Sheet1");
+            ws.Range("C3:D7").AddConditionalFormat();
+            ws.Range("C3:D3").Clear(XLClearOptions.Formats);
+
+            Assert.AreEqual(1, ws.ConditionalFormats.Count());
+            Assert.AreEqual("C4:D7", ws.ConditionalFormats.Single().Range.RangeAddress.ToStringRelative());
+        }
+
+        [Test]
+        public void ClearConditionalFormattingsWhenRangeBellow1()
```
Comment (b0bi79) [inline ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs]: Sorry for my english

**Comment by b0bi79** [comment]
Yes, now I see.

File: ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs (hunk 1 above)
Comment (igitur) [inline ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs]: @b0bi79 No problem. Your English is better than my Russian :)

File: ClosedXML/Extensions.cs [hunk 2]
```diff
@@ -354,4 +359,41 @@ public static bool IsNumber(this object value)
                     || value is decimal;
         }
     }
+
+    public static class RangeExtensions
```
Comment (b0bi79) [inline ClosedXML/Extensions.cs]: I did not want to change the interface IXLRangeBase. Do you think it's better to add new methods into interface?

File: ClosedXML/Excel/Ranges/XLRangeBase.cs [hunk 3]
```diff
@@ -1179,14 +1231,18 @@ private IXLRangeColumns InsertColumnsBeforeInternal(Boolean onlyUsedCells, Int32
...
+                        using (var row = rangeToReturn.Row(ro))
+                            row.Style = styleToUse;
                     }
 
                 }
             }
 
             if (nullReturn)
+            {
+                rangeToReturn.Dispose();
```
Comment (b0bi79) [inline ClosedXML/Excel/Ranges/XLRangeBase.cs]: RangeToReturn is returned from the method. Therefore, we must Dispose it only if nullReturn.

File: ClosedXML/Excel/Ranges/XLRangeBase.cs [hunk 4]
```diff
@@ -1414,14 +1471,18 @@ private IXLRangeRows InsertRowsAboveInternal(Boolean onlyUsedCells, Int32 number
...
+                        using (var column = rangeToReturn.Column(co))
+                            column.Style = styleToUse;
                     }
                 }
             }
 
             // Skip calling .Rows() for performance reasons if required.
             if (nullReturn)
+            {
+                rangeToReturn.Dispose();
```
Comment (b0bi79) [inline ClosedXML/Excel/Ranges/XLRangeBase.cs]: RangeToReturn is returned from the method. Therefore, we must Dispose it only if nullReturn.

File: ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs (hunk 1 above)
Comment (b0bi79) [inline ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs]: Done

File: ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs [hunk 5]
```diff
@@ -170,6 +170,95 @@ public void WsNamedRangesOneString()
...
+            var ws = new XLWorkbook().Worksheets.Add("Sheet1");
+            ws.Range("C3:D7").AddConditionalFormat();
+            ws.Range("B7:E8").Clear(XLClearOptions.Formats);
+
+            Assert.AreEqual(1, ws.ConditionalFormats.Count());
+            Assert.AreEqual("C3:D6", ws.ConditionalFormats.Single().Range.RangeAddress.ToStringRelative());
+        }
+
+        [Test]
+        public void ClearConditionalFormattingsWhenRangeBellow2()
```
Comment (b0bi79) [inline ClosedXML_Tests/Excel/Ranges/XLRangeBaseTests.cs]: Done

File: ClosedXML/Excel/Ranges/XLRangeBase.cs [hunk 6]
```diff
@@ -1542,13 +1625,13 @@ protected void ShiftColumns(IXLRangeAddress thisRangeAddress, XLRange shiftedRan
             if (thisRangeAddress.IsInvalid || shiftedRange.RangeAddress.IsInvalid) return;
 
             if ((columnsShifted < 0
-                 // all columns
+                // all columns
```
Comment (b0bi79) [inline ClosedXML/Excel/Ranges/XLRangeBase.cs]: Done

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 7]
```diff
@@ -2236,6 +2290,7 @@ internal void ShiftFormulaRows(XLRange shiftedRange, int rowsShifted)
                             }
                             else
                                 sb.Append(matchString);
+                            matchRange.Dispose();
```
Comment (b0bi79) [inline ClosedXML/Excel/Cells/XLCell.cs]: Done

File: ClosedXML/Excel/Cells/XLCell.cs [hunk 8]
```diff
@@ -2475,6 +2530,7 @@ internal void ShiftFormulaColumns(XLRange shiftedRange, int columnsShifted)
...
```
Comment (b0bi79) [inline ClosedXML/Excel/Cells/XLCell.cs]: Done

**Comment by b0bi79** [comment]
Francois, please review my PR. I'm developing an open source Excel Template Engine library that uses ClosedXML and I can not upload it to git while ClosedXML does not have these changes.

**Comment by b0bi79** [comment]
I'll come from the other side

**Pull request closed without merging**